### ISSUE TITLE
Null ClientId/ClientName in LogoutRequest (Using SAML2P)

### ISSUE THREAD

**Issue by HiwackVDP** [IdentityServer/IdentityServer4#2751]
Hello all,
I'm using the IdentityServer4 with the QuickStart on asp.net core.
I've actually successfully developped a Saml SSO Idp. The Sign-on part works perfectly.
But for the log-out I need to get the ClientId or ClientName in the Logout action controller.
So I'm trying to use IIdentityServerInteractionService which is injected in the controller.

` LogoutRequest logoutRequest = _interaction.GetLogoutContextAsync(logoutId).Result;`

But here is what I get in logoutRequest : 
![logoutrequest](https://user-images.githubusercontent.com/36201002/47569365-ba7b4380-d933-11e8-92d0-95b24805068b.png)
I know ClientIds contains the ClientId i'm looking for, but as it also takes all the other connected SP Ids in, I can't use it.

I've used Fiddler to look at what is happening behind all of this, and it looks like there isn't any Saml in the logout request the SP send, am I wrong ?

I've seen many people who've successfully used the LogoutRequest, but they mostly use oidc.
So I presume there as to be with a bad configuration on my side, or maybe Saml2 is inappropriate for this ?

Connection to my database is fine, Clients work perfectly for the SSO part as I've said before.
I'm actually quite new to asp.net core and all of this so sorry if this looks confused.

Here is my Idsrv4 config in Startup : 

```
 services.AddDbContext<SamlConfigurationDbContext>(db =>
                        db.UseSqlServer(Configuration.GetConnectionString("ConnectionString"),
                            sql => sql.MigrationsAssembly(migrationsAssembly)));
            services.AddScoped<ISamlConfigurationDbContext, SamlConfigurationDbContext>();


            services.AddIdentityServer()
                .AddSigningCredential(certificate)
                .AddConfigurationStore(options =>
                {
                    options.ConfigureDbContext = builder =>
                        builder.UseSqlServer(Configuration.GetConnectionString("ConnectionString"),
                            sql => sql.MigrationsAssembly(migrationsAssembly));
                })
                .AddOperationalStore(options =>
                {
                    options.ConfigureDbContext = builder =>
                        builder.UseSqlServer(Configuration.GetConnectionString("ConnectionString"),
                            sql => sql.MigrationsAssembly(migrationsAssembly));
                })
                .AddSamlPlugin(ConfigureSaml)
                .AddServiceProviderStore<ServiceProviderStore>()
                .AddProfileService<CustomProfileService>();
```

So in summary here is my question : Why can't I get the ClientId/Name ? and how can I get it ?

**Comment by scottbrady91** [comment]
Just seen this one on Stack Overflow. What version of the SAML library are you using? Seems fine on the latest.

**Comment by scottbrady91** [comment]
Just to link the two lines of communication: https://stackoverflow.com/questions/53062513/net-core-identityserver4-null-clientid-clientname-in-logoutrequest-using-saml2

**Comment by jamieide** [comment]
I'm facing the same problem and I'm using OIDC, not SAML. I have replicated the source code that creates LogoutMessage from logoutId, and ClientId, ClientName, and PostLogoutRedirectUri are missing but the client is in the ClientIds collection. I can't figure out where logoutId comes from so I'm stuck. I am using ASP.NET Core Identity so I'm probably missing a step

**Comment by HiwackVDP** [comment]
@scottbrady91 As I was in formation the 3 last weeks I did not look in precision. But for now nothing new. I can't find anything that would be related to my issue on [the QuickStart you linked on StackOverflow](https://github.com/RockSolidKnowledge/Samples.IdentityServer4.Saml2pIntegration/blob/master/idpWithEf/Startup.cs).

**Comment by jamieide** [comment]
@HiwackVDP I found my problem with oidc configuration, the setting SaveTokens must be set to true.
`options.SaveTokens = true;`
I'm not familiar with SAML configuration but I hope this helps. I'm not sure if this is a bug or not.

**Comment by HiwackVDP** [comment]
Thanks, no problem, we've actually put this feature as a could for now as it's not a key one. Thanks for your support.

**Comment by brockallen** [comment]
Can we close, since this is RSK related?

**Comment by HiwackVDP** [comment]
After nearly a year, we still have this issue.
I can now tell that it's not specific to Saml2P. I've tried with a fresh project using OpenID Connect and the latest Quickstart version.
It's based on this : https://www.scottbrady91.com/Identity-Server/Getting-Started-with-IdentityServer-4

I still get exactly the same result.
Both my client and Idp projects are only composed of what is necessary to test OpenID Connect authentication.

**Comment by brockallen** [comment]
For the oidc scenario, check the logs in the end session request endpoint. This should help explain why you have no context in the logout page.

**Comment by HiwackVDP** [comment]
I can't access the project right now. I'm pretty sure we already gave a try to identityserver verbose logs, but just in case I'll double check tomorrow morning.

**Comment by HiwackVDP** [comment]
Here are the logs I get at signout :

```
2019-12-11 11:04:40.528 +01:00 [DBG] Request path /connect/endsession matched to endpoint type Endsession
2019-12-11 11:04:40.529 +01:00 [DBG] Endpoint enabled: Endsession, successfully created handler: IdentityServer4.Endpoints.EndSessionEndpoint
2019-12-11 11:04:40.529 +01:00 [INF] Invoking IdentityServer endpoint: IdentityServer4.Endpoints.EndSessionEndpoint for /connect/endsession
2019-12-11 11:04:40.530 +01:00 [DBG] Processing signout request for DOMAIN\user
2019-12-11 11:04:40.532 +01:00 [DBG] Start end session request validation
2019-12-11 11:04:40.534 +01:00 [INF] End session request validation success
{"ClientId":null,"ClientName":null,"SubjectId":"DOMAIN\user","PostLogOutUri":null,"State":null,"Raw":{"post_logout_redirect_uri":"http://localhost:59780/signout-callback-oidc","state":"CfDJ8EIIoLaF6m9Aq_X7HQDZvPl0_y7PASOJRkdMmlOnqtxDLdziAQxgWnttS2HEX3lQXccIhRUvo-krJti_59qvAPdEtajwSyoiFCjBQWDzr6Bqtdf0ls6N_sNntnD5WsEEs8BeglwLRdvQrQp4SUliwYmkfrhcgU3bXvQ-NsRXPLl3","x-client-SKU":"ID_NETSTANDARD2_0","x-client-ver":"5.5.0.0"},"$type":"EndSessionRequestValidationLog"}
2019-12-11 11:04:40.534 +01:00 [DBG] Success validating end session request from null
2019-12-11 11:04:40.534 +01:00 [VRB] Invoking result: IdentityServer4.Endpoints.Results.EndSessionResult
2019-12-11 11:04:40.549 +01:00 [VRB] No endpoint entry found for request path: /Account/Logout
2019-12-11 11:04:40.549 +01:00 [VRB] No endpoint entry found for request path: /Account/Logout
2019-12-11 11:04:40.577 +01:00 [INF] {"SubjectId":"DOMAIN\user","DisplayName":"Test User","Category":"Authentication","Name":"User Logout Success","EventType":"Success","Id":1002,"Message":null,"ActivityId":"80000025-0001-fe00-b63f-84710c7967bb","TimeStamp":"2019-12-11T10:04:40.0000000Z","ProcessId":28092,"LocalIpAddress":"127.0.0.1:5001","RemoteIpAddress":"127.0.0.1","$type":"UserLogoutSuccessEvent"}
2019-12-11 11:04:40.650 +01:00 [VRB] No endpoint entry found for request path: /js/signout-redirect.js
2019-12-11 11:04:40.650 +01:00 [VRB] No endpoint entry found for request path: /js/signout-redirect.js
2019-12-11 11:04:40.650 +01:00 [DBG] Request path /connect/endsession/callback matched to endpoint type Endsession
2019-12-11 11:04:40.656 +01:00 [DBG] Endpoint enabled: Endsession, successfully created handler: IdentityServer4.Endpoints.EndSessionCallbackEndpoint
2019-12-11 11:04:40.656 +01:00 [INF] Invoking IdentityServer endpoint: IdentityServer4.Endpoints.EndSessionCallbackEndpoint for /connect/endsession/callback
2019-12-11 11:04:40.659 +01:00 [DBG] Processing signout callback request
2019-12-11 11:04:40.666 +01:00 [VRB] Calling into client configuration validator: IdentityServer4.Saml.Validation.SamlClientConfigurationValidator
2019-12-11 11:04:40.666 +01:00 [DBG] client configuration validation for client openIdConnectClient succeeded.
2019-12-11 11:04:40.666 +01:00 [DBG] No client front-channel logout URLs
2019-12-11 11:04:40.666 +01:00 [DBG] No client back-channel logout URLs
2019-12-11 11:04:40.666 +01:00 [INF] Successful signout callback.
2019-12-11 11:04:40.667 +01:00 [DBG] No client front-channel iframe urls
2019-12-11 11:04:40.667 +01:00 [DBG] No client back-channel iframe urls
2019-12-11 11:04:40.667 +01:00 [VRB] Invoking result: IdentityServer4.Endpoints.Results.EndSessionCallbackResult
```

But I'm not seeing anything of interest in here except the `End session request validation success
{"ClientId":null,"ClientName":null,"SubjectId":"DOMAIN\user","PostLogOutUri":null,"State":null,"Raw":...` part.
Does that mean the client isn't sending enough information ?

**Comment by brockallen** [comment]
>  Does that mean the client isn't sending enough information ?

Perhaps? Check the docs: http://docs.identityserver.io/en/latest/endpoints/endsession.html

**Comment by HiwackVDP** [comment]
It looks like the `HttpContext.SignOutAsync("oidc")` method doesn't include the `id_token_hint` by default in the query string parameters.

So with this code I can successfully send a complete signout request from client side : 
```
        public async Task Logout()
        {
            var token = await HttpContext.GetTokenAsync("oidc", "id_token");
            var props = new AuthenticationProperties
            {
                Parameters =
                {
                    { "id_token_hint", token }
                }
            };

            await HttpContext.SignOutAsync("oidc", props);
        }
```

So this part is completely out of my hands as I'm the server.
I'll check in the days to come if there are the same kinds of (missing) necessary parameters in the SAML logout requests we are receiving from Service Providers.
